pkg/clusteragent/clusterchecks/dispatcher_ksm.go

```diff
@@ -0,0 +1,200 @@
+
+	for _, cfg := range shardedConfigs {
+		patchedCfg, err := d.patchConfiguration(cfg)
+		if err != nil {
+			log.Warnf("Cannot patch resource-sharded KSM config %s: %s", cfg.Digest(), err)
+			continue
+		}
+
+		// Add the shard (will go to dangling if no runners available)
+		d.add(patchedCfg)
+		shardedDigests = append(shardedDigests, patchedCfg.Digest())
+	}
+
+	if len(shardedDigests) == 0 {
+		log.Errorf("KSM sharding enabled but failed to create any shards - falling back to normal scheduling")
+		return false
+	}
+
+	log.Infof("Successfully sharded KSM check into %d resource-grouped checks", len(shardedDigests))
+
+	// Store that we've sharded this config to avoid re-sharding
+	d.markAsSharded(config, shardedDigests)
+
+	return true
+}
+
+// isAlreadySharded checks if a KSM config was already sharded
+// Thread-safe: protected by ksmShardingMutex
+func (d *dispatcher) isAlreadySharded(config integration.Config) bool {
+	d.ksmShardingMutex.Lock()
+	defer d.ksmShardingMutex.Unlock()
+
+	if d.ksmShardedConfig.Name != "" && d.ksmShardedConfig.Digest() == config.Digest() {
+		return true
+	}
+	return false
+}
+
+// markAsSharded marks a KSM config as having been sharded and tracks the sharded digests
+// Thread-safe: protected by ksmShardingMutex
+func (d *dispatcher) markAsSharded(config integration.Config, shardedDigests []string) {
+	d.ksmShardingMutex.Lock()
+	defer d.ksmShardingMutex.Unlock()
+
+	d.ksmShardedConfig = config
+	d.ksmShardedDigests = shardedDigests
+}
+
+// unscheduleKSMCheck removes all sharded KSM configs if this is a sharded KSM check
+// Returns true if sharded configs were removed, false otherwise
+// Thread-safe: protected by ksmShardingMutex
+func (d *dispatcher) unscheduleKSMCheck(config integration.Config) bool {
+	if !d.ksmSharding.IsEnabled() || !d.ksmSharding.IsKSMCheck(config) {
+		return false
+	}
+
+	if !d.isAlreadySharded(config) {
+		return false
+	}
+
+	// Atomically read the digests and clear tracking
+	// We must hold the lock to avoid TOCTOU race between checking and reading digests
+	d.ksmShardingMutex.Lock()
+	digestsCopy := make([]string, len(d.ksmShardedDigests))
+	copy(digestsCopy, d.ksmShardedDigests)
+	digestCount := len(digestsCopy)
+
+	// Clear the tracking while holding the lock
+	d.ksmShardedConfig = integration.Config{}
+	d.ksmShardedDigests = nil
+	d.ksmShardingMutex.Unlock()
+
+	log.Infof("Unscheduling sharded KSM check %s (removing %d shards)", config.Digest(), digestCount)
+
+	// Remove all sharded configs (outside the lock since this is expensive)
+	for _, digest := range digestsCopy {
+		log.Debugf("Removing KSM shard with digest %s", digest)
+		d.removeConfig(digest)
+	}
+
+	log.Infof("Successfully unscheduled all sharded KSM configs")
+	return true
+}
+
+// getAvailableRunners returns list of available cluster check runner node names
+// Only returns nodes with NodeTypeCLCRunner, filtering out NodeTypeNodeAgent
+func (d *dispatcher) getAvailableRunners() []string {
```

Comment:
I think this can be simplified.
The only caller of this function just needs the number of runners